[stdlib] Revising Array Internal Comments and Example Code for Consistency.

devKobe24 opened the pull request:
In the documentation from line 92 to 96, there is a statement that reads, "Using a negative number or an index equal to or greater than count triggers a runtime error."

I believe that a more accurate example is needed for this statement.

Accordingly, I have modified the existing example code so that the example code and the explanatory comments are consistent.

With these modifications, users reading the comments can understand more clearly, and it is anticipated that they won't have to go through the process of looking up additional documentation.

Previous code:
```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
```

Modified code:
```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(oddNumbers[-1])
///	// Using a negative number
///	// Triggers runtime error: Index out of range
///
///     print(oddNumbers[8])
///     // Equal to `count`
///     // Triggers runtime error: Index out of range
///
///     print(oddNumbers[9])
///     // Greater than `count`
///     // Triggers runtime error: Index out of range
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
```

devKobe24 committed:
[stdlib] Revising Array Internal Comments and Example Code for Consistency

In the documentation from line 92 to 96, there is a statement that reads, "Using a negative number or an index equal to or greater than count triggers a runtime error."

I believe that a more accurate example is needed for this statement.

Accordingly, I have modified the existing example code so that the example code and the explanatory comments are consistent.

With these modifications, users reading the comments can understand more clearly, and it is anticipated that they won't have to go through the process of looking up additional documentation.

Previous code:
```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
```

Modified code:
```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(oddNumbers[-1])
///		// Using a negative number
///		// Triggers runtime error: Index out of range
///
///     print(oddNumbers[8])
///     // Equal to `count`
///     // Triggers runtime error: Index out of range
///
///     print(oddNumbers[9])
///     // Greater than `count`
///     // Triggers runtime error: Index out of range
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
```

devKobe24 committed:
Merge branch 'apple:main' into develop

devKobe24 commented:
@swift-ci please smoke test

devKobe24 commented:
Whenever you have time, I would appreciate it if you could provide a review. Thank you.
cc: @invalidname

devKobe24 commented:
Could it be that the review is not proceeding due to a mistake on my part? If there is anything wrong, please let me know, and I will make the necessary corrections.

cc: @glessard

invalidname commented:
> Could it be that the review is not proceeding due to a mistake on my part? If there is anything wrong, please let me know, and I will make the necessary corrections.

You opened the request on a three-day weekend in the US (Martin Luther King Jr Day), so many of us haven’t been back to work since then.

I’ll have a look in the morning.

invalidname reviewed (dismissed):
The paragraph above describes these cases as throwing a runtime error, but it doesn't hurt to spell it out explicitly like this. Maybe this helps the new Swift developer who hasn't been using arrays in various languages for many years.

amartini51 commented on stdlib/public/core/Array.swift:
```diff
@@ -98,6 +98,18 @@
 ///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
 ///     // Prints "1, 7"
 ///
+///     print(oddNumbers[-1])
+///     // Using a negative number
+///     // Triggers runtime error: Index out of range
+///
+///     print(oddNumbers[8])
+///     // Equal to `count`
+///     // Triggers runtime error: Index out of range
+///
+///     print(oddNumbers[9])
+///     // Greater than `count`
+///     // Triggers runtime error: Index out of range
+///
```
Instead of adding a code example of every way to use an array incorrectly, maybe rewriting the paragraph above would help?  The prose already describes these invalid conditions, but it might be easier to read as a bulleted list.

Repeating the list from the paragraph in code makes the listing pretty long, and might actually make this harder to read.

devKobe24 replied:
As a user of Swift, I would like to share my personal experience: I have found myself perplexed on occasions due to the lack of examples accompanying the explanations. In the case of Array, for instance, although the comments are neatly organized, the absence of detailed explanations means users may need to go through the trouble of referring to Apple's official documentation. If explanations were included, I believe it would not only facilitate a quicker understanding of the language but also offer easier accessibility, combining the benefits of both rapid comprehension and user-friendliness

amartini51 replied:
The comments that begin with `///` _are_ the documentation.  Any text we write here appears in the API reference for `Array`.  For example, the content this PR suggests changing is published here:

https://developer.apple.com/documentation/swift/array#Accessing-Array-Values

devKobe24 replied:
Previously, there was an example about Arrays in the comments as well. I added a more detailed example to those comments with the intention of facilitating a better understanding. As I mentioned earlier, another advantage of this is that it can save the effort of consulting the official Apple documentation. That's why I included the additional example as described above.

devKobe24 replied:
I would like to inquire whether it would be a good idea to delete the code from lines 101 to 112.

As a user, I often utilize the 'Jump to Definition' feature in Xcode when I want to learn about how to use a function or am curious about its internal implementation while developing an app or working on a project. When I use this feature, I find that the comments not only explain the internal workings but also provide detailed instructions on how to use the functions, allowing me to understand them thoroughly without needing to consult Apple's official documentation. Therefore, based on my experience, I suggest that adding examples to the existing comments explaining arrays in the Pull Request would make the descriptions longer but would significantly enhance the understanding and clarity of arrays. I believe this user-friendly approach would greatly benefit others.

cc: @amartini51

amartini51 reviewed (changes requested):
Thanks for this suggested change! I think this approach has some issues, and described an alternate approach in comments.

devKobe24 commented:
As a user, I often utilize the 'Jump to Definition' feature in Xcode when I want to learn about how to use a function or am curious about its internal implementation while developing an app or working on a project. When I use this feature, I find that the comments not only explain the internal workings but also provide detailed instructions on how to use the functions, allowing me to understand them thoroughly without needing to consult Apple's official documentation. Therefore, based on my experience, I suggest that adding examples to the existing comments explaining arrays in the Pull Request would make the descriptions longer but would significantly enhance the understanding and clarity of arrays. I believe this user-friendly approach would greatly benefit others.

cc: @invalidname @amartini51

devKobe24 committed:
[chore] Restore comments inside the Array.swift

In accordance with your requested changes, I have respectfully removed the comments on lines 101 to 112.

devKobe24 committed:
Merge branch 'apple:main' into develop

devKobe24 commented:
In accordance with your requested changes, I have respectfully removed the comments on lines 101 to 112.

cc: @invalidname @amartini51

devKobe24 commented:
Have I made any mistakes in my actions or comments throughout the process that have delayed the review? 
Please let me know, and I will address the issue as quickly as possible. 
Thank you.

cc: @glessard

amartini51 commented:
> In accordance with your requested changes, I have respectfully removed the comments on lines 101 to 112.

Thanks.  With that change, this PR no longer makes any changes.  Would you like to try turning the prose into a bulleted list, as discussed in the comments?  I'll try to make a PR next week with that change, if time permits.

devKobe24 committed:
[stdlib] Creating Bullet Point Lists Within an Array for Consistency.

The comments from line 92 to 96 mention that using an index that is negative or greater than the count will cause a runtime error.

I believe we need a more precise example for this comment.

Accordingly, I have made edits to ensure that the example code and the explanatory comments are consistent, using bullet points for clarity and coherence.

This should allow readers to understand the content more clearly and consistently, reducing the need to search for additional documentation.

Previous Code:

```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
```

Modified Code:

```swift
/// You can access individual array elements through a subscript. The first
/// element of a nonempty array is always at index zero. You can subscript an
/// array with any integer from zero up to, but not including, the count of
/// the array. Using a negative number or an index equal to or greater than
/// `count` triggers a runtime error. For example:
///
///     print(oddNumbers[0], oddNumbers[3], separator: ", ")
///     // Prints "1, 7"
///
///     print(emptyDoubles[0])
///     // Triggers runtime error: Index out of range
///
///     - print(oddNumbers[-1])
///     	- Using a negative number
///     	- Triggers runtime error: Index out of range
///
///     - print(oddNumbers[8])
///     	- Equal to `count`
///     	- Triggers runtime error: Index out of range
///
///     - print(oddNumbers[9])
///     	- Greater than `count`
///     	- Triggers runtime error: Index out of range
```

devKobe24 committed:
Merge branch 'apple:main' into develop

devKobe24 commented:
> Thanks. With that change, this PR no longer makes any changes. Would you like to try turning the prose into a bulleted list, as discussed in the comments? I'll try to make a PR next week with that change, if time permits.
> 
> I think the history might be cleaner if we close this PR without merging, and open a new one to try doing this a different way. This branch picked up several merge commits that aren't needed.

Thank you for your kind response.
I apologize as I sent the PR again without fully understanding your answer.
I will send another PR once it is created next week.
I will close this PR for now.

amartini51 commented:
Here's an example of using a bulleted list:

https://github.com/apple/swift/compare/main...amartini51:swift:array_traps

It still merits discussion whether this is actually an improvement.

devKobe24 commented:
> Here's an example of using a bulleted list:
> 
> [main...amartini51:swift:array_traps](https://github.com/apple/swift/compare/main...amartini51:swift:array_traps)
> 
> It still merits discussion whether this is actually an improvement.

Allow me to share my personal opinion: I understand Swift to be a user-friendly language, a sentiment I have experienced firsthand.

Consequently, even for someone new to Swift, the 'Jump to definition' feature provides clear, user-friendly examples that facilitate immediate comprehension through direct engagement.

Additionally, this reduces the need to search for supplementary documentation, swiftly bringing users to appreciate another of Swift's strengths: its flexibility.

Lastly, reading comments can open up a vista for exploring deeper knowledge. I, too, have often stumbled upon other texts while reading well-explained, user-friendly examples in comments, leading to the acquisition of new knowledge.

For these reasons, I believe there are more positive aspects from a user's perspective.
